### DISCUSSION TITLE
How to register $client as a global variable and support typescript in Nuxt

### DISCUSSION THREAD

**Discussion by StringKe** [Colonel-Sandvich/trpc-vue-query#16] [General]
I'd like the more convenient way, rather than through the useNuxtApp composables.

**Comment by Colonel-Sandvich** [comment]
So you can just [create the trpc client anywhere]( https://trpc.io/docs/client/vanilla/setup#3-initialize-the-trpc-client) and then import it anywhere. And that will be like your global variable.

Then for trpc-vue-query you can do
```ts
import { client } from './your-path-to-trpc'

createTrpcVueClient(client, useQueryClient())
```

As for supporting typescript in Nuxt I'm not sure what you're asking for as I have typescript working in multiple Nuxt projects. If you're IDE doesn't have typescript working it could be related to the Vue (Volar) extension upgrading to V2. Check out [their issues](https://github.com/vuejs/language-tools/issues)

**Reply by StringKe** [reply]
I tried to do this, but nuxt gives me an error.


~/utils/api.ts
```
import { createTrpcVueClient, customFetchWrapper } from '@colonel-sandvich/trpc-vue-query';
import { useQueryClient } from '@tanstack/vue-query';
import { httpBatchLink } from '@trpc/client';

import type { AppRouter } from '~/server/routers';

export const apiClient = createTrpcVueClient<AppRouter>(
    {
        links: [
            httpBatchLink({
                url: '/api/trpc',
                headers: useRequestHeaders(),
                fetch: customFetchWrapper(),
            }),
        ],
    },
    useQueryClient(),
);
```

> [nuxt] A composable that requires access to the Nuxt instance was called outside of a plugin, Nuxt hook, Nuxt middleware, or Vue setup function. This is probably not a Nuxt bug. Find out more at `https://nuxt.com/docs/guide/concepts/auto-imports#vue-and-nuxt-composables`.

**Comment by Colonel-Sandvich** [comment]
So you can't create the trpcVueClient outside of the Nuxt context, because @tanstack/vue-query needs access to that context. You should create it in a nuxt plugin. If however you need access to trpc's client then you can create that in ~/utils/api.ts and then import it and pass it to `createTrpcVueClient` like my code above suggests.

**Reply by StringKe** [reply]
Can you give me an example? The client has been created using `createTrpcVueClient` in the plugin, is it still `createTrpcVueClient` created in utils/api?

I checked the type definition of createTrpcVueClient and he doesn't support passing a client.

**Reply by Colonel-Sandvich** [reply]
```ts
// ~/utils/trpcClient.ts

import { createTRPCClient, httpBatchLink } from '@trpc/client';
import type { AppRouter } from '../path/to/server/trpc';
export const client = createTRPCClient<AppRouter>({
  ...
});
```

```ts
// ~/plugins/augmentedClient.ts
import { client } from "~/utils/trpcClient"

export default defineNuxtPlugin({
  name: "myClient",
  dependsOn: ["vueQuery"],
  setup: () => {
    const client = createTrpcVueClient(trpc, useQueryClient());

    return {
      provide: {
        client,
      },
    };
  },
});
```

So the trpcClient in utils you can use anywhere because it does not need Nuxt context. But the client from /plugins/ does need Nuxt context to be used.

And yes, as the creator of this package, I am sure that `createTrpcVueClient` supports passing a client. Make sure to use the latest version of the package.
